Dev collection.bootstrap dnf

pvtorres opened the pull request:
Updates to power_aix_bootstrap role to allow dnf, wget and pycurl installations. 
Background: 
AIX 7.3 comes with python3 per-installed. Ansible will work with a fresh installed or migrated 7.3 LPAR without the need to install yum and python. 

If any open-source from the AIX toolsbox is needed,  dnf should be installed in AIX 7.3 
This pull request will add features to the power_aix_bootstrap role to automatically install dnf, wget (required by the flrtvc module).

How to use these new features:
New demo playbook:
demo_boostrap_dnf.yml:



---
- name: "Bootstrap dnf on AIX"
  hosts: aix
  remote_user: root
  gather_facts: no
  collections:
  - ibm.power_aix
  tasks:

\# CHECK for dnf on inventory host
  - import_role:
      name: power_aix_bootstrap
    vars:
      pkgtype: "dnf"
      download_dir: "~"
      target_dir: "/tmp/.ansible.cpdir"

\# INSTALL / UPDATE wget
- name: "Bootstrap wget"
  hosts: aix
  remote_user: root
  gather_facts: no
  collections:
  - ibm.power_aix
  tasks:

  - import_role:
      name: power_aix_bootstrap
    vars:
      pkgtype: "wget"
\#     opt_size: 500

\# INSTALL / UPDATE pycurl
- name: "Bootstrap pycurl"
  hosts: aix
  remote_user: root
  gather_facts: no
  collections:
  - ibm.power_aix

  tasks:
  - import_role:
      name: power_aix_bootstrap
    vars:
      pkgtype: "pycurl"
\#     opt_size: 500




Considerations: 
*  dnf is installed by the AIX toolsbox in the /opt/freeware/bin folder, not in /usr/bin 
* dnf requires the use of the python interpreter installed with the AIX toolsbox at /opt/freeware/bin/python3
*  playbooks using dnf will need to set the environment option with PATH:/opt/freeware/bin so the Ansible dnf module 
can locate the dnf binary. 
* playbooks using the dnf module will have to also set the ansible_python_interpreter to /opt/freeware/bin/python3. 
* The new bootstrap features, dnf, wget, pycurl, will take care of the PATH, interpreter and any other dependencies. 


Example of using the dnf module to install open-source tools. Notice the set of the ansible_python_interpreter and the PATH variables.

---
- name: "Install wget using dnf"
  hosts: all
  user: root
  gather_facts: no
  vars:
    ansible_python_interpreter: /opt/freeware/bin/python3
    PATH: /opt/freeware/bin

  collections:
  - ibm.power_aix

  tasks:
  - name: Check for wget command
    dnf:
      name: "wget"
      state: latest

Example of using the wget bootstrap to install wget. 
The role will take care of handling the interpreter and PATH variables. Also this role will use yum or dnf depending of what it is installed in the system.


---
- name: "FLRTVC on AIX playbook"
  hosts: all
  gather_facts: no
  vars:
    apar: sec
    path: /usr/sys/inst.images
    check: no
    download_only: yes
    extend: yes

  collections:
  - ibm.power_aix

  tasks:
  - import_role:
      name: power_aix_bootstrap
    vars:
      pkgtype: "wget"

  - name: "FLRTVC"
    flrtvc:
      path: "{{ path }}"
      verbose: yes
      apar: "{{ apar }}"
      force: no
      clean: False
      check_only: "{{ check }}"
      download_only: "{{ download_only }}"
      extend_fs: "{{ extend }}"
    register: result
  - debug: var=result




Development insights
-----------------------
power_aix_bootstrap structure before:

<img width="266" alt="image" src="https://user-images.githubusercontent.com/81318963/142081134-16516975-db3e-40c1-8e2d-e5212dfa26d6.png">


power_aix_bootstrap structure after:

<img width="294" alt="image" src="https://user-images.githubusercontent.com/81318963/142081202-c581655d-0fa3-4e93-86c1-519c80013a63.png">


Specifications:

* The bootstrap will dnf will not install dnf in AIX version 7.1 and 72. Use the yum boostrap instead. 
  dnf will only work with python3 32 bit version in AIX 7.1 and 7.2 
* The bootstrap will not install yum in AIX 7.3. Yum does not support python3. 
* The wget bootstrap can install wget either using yum or dnf. 
*  Users can skip bootstraping using the collection and they can just download the AIX toolsbox from the web and take care of the requirements themselves.

pvtorres committed:
First prototype. Bootstrap dnf, wget, pycurl and demo playbooks

pvtorres committed:
update flrtvc playbook with the wget bootstrap to be able to handle 7.3 or 7.2 hosts.

pvtorres committed:
prevent bootstrapping yum on 7.3.
modify the nim flrtvc to install wget if needed.

pvtorres committed:
Uncomment demo bootstrap wget and pycurl.

Clean up for the dnf_aixtoolbox.sh script

pvtorres committed:
ansible-lint --force-color -f pep8 <role> clean up

pvtorres committed:
Remove most of the warning from yum_installer.yml after using ansible-lint

jdejoya17 commented on roles/power_aix_bootstrap/vars/main.yml:
```diff
@@ -1,5 +1,7 @@
-# Copyright (c) IBM Corporation 2020
-pkgtype: "yum"
+# Copyright (c) IBM Corporation 2021
+pkgtype: ""
```
do you really need `pkgtype` in here. this is a required parameter for the bootstrap role. before the default value
was `yum`, so if you did not enter anythin for the `pkgtype` var for the role it would default to yum.
here you made it empty, which i assume you did not want to make `yum` bootstrap the default behavior.

also, you already have a check in the main.yml if `pkgtype` is not specified, which usually indicates this
is a required parameter. so i think this defining `pkgtype` in var/main.yml is not needed

probably just indicate in the updated README.md for the bootstrap role that `pkgtype` is required

jdejoya17 commented on roles/power_aix_bootstrap/tasks/pycurl_install.yml:
```diff
@@ -0,0 +1,129 @@
+# Copyright (c) IBM Corporation 2021
+---
+# Install pycurl and dependencies
+
```
extra newline

jdejoya17 commented on roles/power_aix_bootstrap/tasks/pycurl_install.yml:
```diff
@@ -0,0 +1,129 @@
+  debug:
+     msg: "End play for host {{ aix_host }} because pycurl is installed."
+  when: pycurl_exists.stdout is search("true")
+
```
extra newline

pvtorres replied:
fixed

jdejoya17 commented on roles/power_aix_bootstrap/tasks/pycurl_install.yml:
```diff
@@ -0,0 +1,129 @@
+       include_tasks: increase_filesystem.yml
+       with_items:
+          - {filesystem: "opt", size_needed: "{{ opt_size }}"}
+          - {filesystem: "var", size_needed: 50}
```
probably have a variable for the size_needed for `/var` filesystem, similar to what you have with `opt_size`

jdejoya17 commented on playbooks/demo_bootstrap_dnf.yml:
```diff
@@ -0,0 +1,47 @@
+      name: power_aix_bootstrap
+    vars:
+      pkgtype: "dnf"
+      download_dir: "~"
```
might want to give the absolute path

pvtorres replied:
download_dir is the path to download the scripts and tar files from the AIX toolsbox. User can change this path as needed.

jdejoya17 commented on roles/power_aix_bootstrap/vars/main.yml:
```diff
@@ -1,5 +1,7 @@
 target_dir: "/tmp/.ansible.cpdir.yum"
 aix_host: "{{ hostvars[inventory_hostname].inventory_hostname }}"
+interpreter_name: /opt/freeware/bin/python3
+opt_size: 500
```
shouldn't this be in `defaults/main.yml` instead, since this is a default value

jdejoya17 commented on playbooks/demo_flrtvc_dnf.yml:
```diff
@@ -0,0 +1,34 @@
+  tasks:
+  - name: Check for wget command
+    dnf:
+      name: "wget"
```
you already have a bootstrap feature to install wget. i think you can use the import role here for bootstrap role

pvtorres replied:
The idea of the demo playbook is to show an example on using the dnf module with the extra variables (interpreter and path).
I thought about the possibility for confusion. What if I just rename the module demo_dnf.yml and show only the installation of wget?

jdejoya17 reviewed (changes requested):
i have added some comments

jdejoya17 committed:
update sanity test workflow and ignore file for ansible 2.13 (#174)

Co-authored-by: Joseph de Joya <Joseph.de.Joya@ibm.com>

pvtorres committed:
Changes to the power_aix_bootstrap due to code reviews
- Delete the demo_fltvc_dnf.yml and add a new demo_dnf.yml
- Move variables from vars to default.
  Variables in var should be variales that are not usually modified.
  default is a better place for variables that are inputs for the role.
- Add additional information for the role's documentation. README.md

MGomathi commented on roles/power_aix_bootstrap/tasks/dnf_install.yml:
```diff
@@ -0,0 +1,166 @@
+          dest: "{{ download_dir }}/{{ dnf_src_71_72 }}"
+          mode: '0644'
+       when: not dnf_71_72_dest.stat.exists
+       delegate_to: localhost
```
We exit if aix version is 7.3 to install dnf. Then do we really need to download dnf for AIX 7.1, 7.2?

jdejoya17 replied:
@pvtorres is this comment resolved?

pvtorres committed:
Minor changes to the power_aix_bootstrap documentation.

MGomathi commented on roles/power_aix_bootstrap/tasks/dnf_install.yml:
```diff
@@ -0,0 +1,166 @@
+
+  when: (output.stdout_lines[0] | int) >= 73
+
+# { Excute this block if dnf is not installed and the AIX version is igual or greater than 7.3
```
Small typo in the comment

pvtorres commented on roles/power_aix_bootstrap/README.md:
```diff
@@ -11,9 +11,32 @@ None.
+
+Specifies the free space in megabytes needed in the /opt folder. It is used by dnf, wget and pycurl bootstraps.
+
+    var_free_size (optional, str, 500)
```
Change 50 MB

pvtorres commented on roles/power_aix_bootstrap/README.md:
```diff
@@ -11,9 +11,32 @@ None.
 Available variables are listed below, along with default values:
 
-    pkgtype (True, str, yum)
+    pkgtype (True, str, none)
```
Find if the True argument can be replace for another more meaningful one and also if none is the right value.

pvtorres replied:
There is not a default way to show which values are required, type or values. 
I added more elaborated description of the arguments, values and descriptions.

jdejoya17 replied:
@pvtorres found anything on this?

pvtorres commented on roles/power_aix_bootstrap/files/dnf_aixtoolbox.sh:
```diff
@@ -0,0 +1,168 @@
+#!/bin/ksh
```
Rename the script to dnf_install.sh

pvtorres commented on roles/power_aix_bootstrap/tasks/dnf_install.yml:
```diff
@@ -0,0 +1,166 @@
+
+     - name: Check for dnf script file on localhost
+       stat:
+          path: "{{ download_dir }}/dnf_aixtoolbox.sh"
```
Use the new name of the script.

pvtorres committed:
Review Fixes
- Readme.md for power_aix_bootstrap modified to reflect var_free_size of 50MB by default.
- Renaming of the dnf_aixtoolbox.sh script to dnf_installer.sh
- Typo fixes.

pvtorres committed:
Review fixes.
Do not remove target folders for yum and  wget if they already existed in the system.
If the filesystem do not exist, the role then will create them and remove them when
finish.

jdejoya17 commented on roles/power_aix_bootstrap/tasks/wget_install.yml:
```diff
@@ -0,0 +1,134 @@
+      register: os_level
+      changed_when: false
+
+    - name: Verify if the operating system is AIX 7.2 and bellow as
```
misspelling `bellow`, should be `below`

jdejoya17 reviewed (commented):
added minor comments

pvtorres committed:
Changes to the power_aix_bootstrap due to review and test.

- Change the location of the PATH and interpreter variables to the dnf and pip modules.
This will prevent the PATH and interpreter to be used by other modules.

- Modify the dnf_install.yml disable any tar bundle downloads
  for AIX 7.1 and AIX 72 because dnf bootstrap is supported only in AIX 7.3

- Modify pycurl_install.yml to install the gcc compiler.
  Also update PIP to prevent problems while installing pycurl. (problem observed
  with the Gold version of AIX 7.3).

- Do not use the the pip module to install pycurl.
Use the shell command to prevent errors. prevent errors.
- Remove the creation  a temporary filesystem for the dnf cache.
  The module just need to increase /var and /opt to have enough space to hold
the cache and the packages installed in the system.

pvtorres committed:
Moving the installation of PIP after the dnf pycurl dependencies are installed.
This was causing and error during the VIOS health test role while calling
the dnf role action then the pycurl one.

jdejoya17 commented on roles/power_aix_bootstrap/tasks/dnf_install.yml:
```diff
@@ -0,0 +1,184 @@
+       register: dnf_73_dest
+       delegate_to: localhost
+
+#     - name: Download dnf bundle file for 7.1 and 7.2 to the controller
```
should this just be deleted? if not used

pvtorres committed:
Update the playbook to run the power_aix_bootstrap pycurl action to use
opt_free_size of 1000 MB.
